perception/autoware_tensorrt_yolox/src/tensorrt_yolox_node.cpp

```diff
@@ -90,6 +91,10 @@ TrtYoloXNode::TrtYoloXNode(const rclcpp::NodeOptions & node_options)
   const tensorrt_common::BatchConfig batch_config{1, 1, 1};
   const size_t max_workspace_size = (1 << 30);
 
+  if (!setCudaDeviceId(gpu_id_)) {
```

Comment:
@manato -san, thank you for your valuable comments. Since I am using the same 3xGPU model for our tests, the created engine file for GPU 0 (default) is valid for the remaining two GPUs. If there is another different model GPU, it will be a problem. Therefore, we can update the [```tensorrt_common```](https://github.com/autowarefoundation/autoware.universe/tree/main/common/tensorrt_common) package to handle different GPUs as well. What do you think about that?

Reply:
@ismetatabay
Thank you for your consideration! 
Yes, my main concern was such kinds of environments that different model GPUs exist. Regarding building engine on the specific device, it seems to be enough to call `cudaSetDevice()` before an engine builder starts to work, according to [FAQs: Q.How do I use TensorRT on multiple GPUs?](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#faq)  (I believe you can confirm this behavior by monitoring which GPU is used using `nvidia-smi` during engine building). If this is the case, I think we don't need to modify `tensorrt_common` because we just need to call `cudaSetDevice` before constructing `tensorrt_common::TrtCommon`.

Sorry for repeatedly asking, but could you consider to move this cuda-related codes into `tensorrt_yolox.cpp` for further encapsulation so that other nodes who use yolox as a module can enjoy your improvement?

Reply:
Thank you @manato -san, you are right. I am okay with moving everything to ```tensorrt_yolox.cpp```. With this update, ```traffic_light_fine_detector``` can also select GPU as you mentioned. However, if I move everything to ```TrtYoloX``` class, I still get an illegal memory access error from CUDA unless I call ```cudaSetDevice``` before creating the ```trt_yolox_``` object in the ```TrtYoloXNode``` constructor. Do you have any suggestions on that?

https://github.com/autowarefoundation/autoware.universe/blob/3e79b8f67f5a3428c04efb7545c57b6401c83d17/perception/autoware_tensorrt_yolox/src/tensorrt_yolox_node.cpp#L92-L97

Reply:
Thank you @ismetatabay -san for your thoughtful consideration!

I guess the cause of the issue you observed is [this](https://github.com/autowarefoundation/autoware.universe/blob/3e79b8f67f5a3428c04efb7545c57b6401c83d17/perception/autoware_tensorrt_yolox/include/autoware/tensorrt_yolox/tensorrt_yolox.hpp#L270) line; the line attempts to create `cudaStream` on the default (`0`) device during construction of an instance, and the stream will be used for memory copy before/after inference. Could you please try to move to call `makeCudaStream()` after setting the device in the class constructor and see if it makes a difference?

Reply:
Hello @manato -san, sorry for the late update. I was on leave and only had a chance to update it today. I have moved the updates to the tensorrt_yolox. Could you please check it? Thank you :raised_hands:

Reply:
Hi @ismetatabay -san, thank you for your continuous updates. The modifications look good to me!